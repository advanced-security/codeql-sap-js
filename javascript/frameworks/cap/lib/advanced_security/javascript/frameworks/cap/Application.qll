/**
 * Definitions pertaining to the application as a whole.
 */

import javascript
import advanced_security.javascript.frameworks.cap.PackageJson

class RootDirectory extends Folder {
  RootDirectory() {
    exists(PackageJson packageJson | this = packageJson.getJsonFile().getParentContainer())
  }

  /**
   * Gets the path of a file relative to this root directory.
   */
  string getFilePathRelativeToRoot(File file) {
    result = file.getAbsolutePath().regexpReplaceAll(this.getAbsolutePath(), ".") and
    result.charAt(0) = "."
  }

  /**
   * Holds if this root directory of the application contains the given file.
   */
  predicate contains(File file) { exists(this.getFilePathRelativeToRoot(file)) }
}
