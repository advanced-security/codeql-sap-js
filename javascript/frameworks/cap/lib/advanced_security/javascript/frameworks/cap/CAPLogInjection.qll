import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    this = any(CdsFacade cds).getMember("log").getACall() and
    name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof TemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"]
    )
  }
}
