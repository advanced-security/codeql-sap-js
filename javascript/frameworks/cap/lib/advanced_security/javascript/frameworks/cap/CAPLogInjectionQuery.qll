import javascript
import semmle.javascript.dataflow.DataFlow
import semmle.javascript.security.dataflow.LogInjectionQuery
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.dataflow.DataFlow

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    exists(CdsFacade cds |
      this = cds.getMember("log").getACall() and
      name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
    )
  }

  string getName() { result = name }
}

class ConstantOnlyTemplateLiteral extends TemplateLiteral {
  ConstantOnlyTemplateLiteral() {
    forall(Expr e | e = this.getAnElement() | e instanceof TemplateElement)
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`.
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"] and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof ConstantOnlyTemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log
    )
  }
}

class CAPLogInjectionConfiguration extends LogInjectionConfiguration {
  CAPLogInjectionConfiguration() { this = "CAP Log Injection" }

  override predicate isSource(DataFlow::Node start) {
    super.isSource(start)
    or
    start instanceof RemoteFlowSource
  }

  override predicate isBarrier(DataFlow::Node node) {
    exists(HandlerParameterData handlerParameterData |
      node = handlerParameterData and
      not handlerParameterData.getType() = ["cds.String", "cds.LargeString"]
    )
  }

  override predicate isSink(DataFlow::Node end) { end instanceof CdsLogSink }
}
