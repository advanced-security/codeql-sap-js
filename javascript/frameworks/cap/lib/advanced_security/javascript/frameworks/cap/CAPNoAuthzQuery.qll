import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.CDL
import advanced_security.javascript.frameworks.cap.Conditionals

abstract class CdlElementWithoutAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class CdlElementWithoutCdsAuthn extends CdlElementWithoutAuthn instanceof CdlElement {
  CdlElementWithoutCdsAuthn() { this.hasNoCdsAccessControl() }
}

class CdlServiceWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlService { }

class CdlEntityWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlEntity {
  CdlEntityWithoutCdsAuthn() {
    this.belongsToServiceWithNoAuthn()
    or
    exists(CdlEntityWithoutCdsAuthn otherCdlEntityWithoutCdsAuthn |
      this.inherits(otherCdlEntityWithoutCdsAuthn)
    )
  }
}

class CdlActionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlActionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlFunctionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlFunctionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlElementProtectionWithHandlerRegistration instanceof HandlerRegistration {
  string toString() { result = super.toString() }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  CdlElementProtectionWithHandlerRegistration() {
    (
      this.isBefore()
      or
      this.isOn()
    ) and
    exists(Handler handler, ConditionalExprOrStatement exprOrStmt |
      handler = this.getHandler() and
      (
        exprOrStmt = handler.getFunction().getBody() or
        exprOrStmt = handler.getFunction().getABodyStmt()
      )
    |
      exprOrStmt.getConditionExpr().getAChildExpr*().flow() instanceof RequestUserIs and
      exists(CdsRequest req |
        exprOrStmt.getPolarity() = true and
        exprOrStmt.getAnElseBranchExpr() = req.getARejectCall().asExpr()
        or
        exprOrStmt.getPolarity() = false and
        exprOrStmt.getAThenBranchExpr() = req.getARejectCall().asExpr()
      )
    )
  }

  string getEntityName() { result = super.getEntityName() }

  string getAnEventName() { result = super.getAnEventName() }
}

abstract class CdlElementWithJsAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

class CdlServiceWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlService {
  CdlServiceWithJsAuthn() {
    exists(CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = "*"
    )
  }
}

class CdlEntityWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlEntity {
  CdlEntityWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnEntity() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getEntityName() = this.getUnqualifiedName()
    )
  }
}

class CdlActionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlAction {
  CdlActionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnAction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

class CdlFunctionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlFunction {
  CdlFunctionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAFunction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

abstract class CdlElementWithoutJsAuthn extends CdlElementWithoutAuthn instanceof CdlElement { }

class CdlServiceWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlService {
  CdlServiceWithoutJsAuthn() { not this instanceof CdlServiceWithJsAuthn }
}

class CdlEntityWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlEntity {
  CdlEntityWithoutJsAuthn() { not this instanceof CdlEntityWithJsAuthn }
}

class CdlActionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlAction {
  CdlActionWithoutJsAuthn() { not this instanceof CdlActionWithJsAuthn }
}

class CdlFunctionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlFunction {
  CdlFunctionWithoutJsAuthn() { not this instanceof CdlFunctionWithJsAuthn }
}

/**
 * The access to property `user` of a handler's request.
 */
class RequestUser extends SourceNode instanceof PropRef {
  RequestUser() {
    exists(Handler handler |
      this.getBase().getALocalSource() = handler.getRequest() and
      this.getPropertyName() = "user"
    )
  }
}

class RequestUserIs instanceof MethodCallNode {
  string userRole;

  RequestUserIs() {
    exists(RequestUser requestUser |
      this = requestUser.getAMethodCall("is") and
      userRole = this.getArgument(0).getStringValue()
    )
  }

  string toString() { result = super.toString() }
}
