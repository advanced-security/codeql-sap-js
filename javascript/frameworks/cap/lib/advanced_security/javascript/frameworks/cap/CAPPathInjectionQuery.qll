/**
 * Exported functions from CAP `cds.utils`.
 * Functions described from:
 * https://www.npmjs.com/package/@sap/cds?activeTab=code
 */

import javascript
import advanced_security.javascript.frameworks.cap.CDSUtils

abstract class UtilsSink extends DataFlow::Node {
  abstract string sinkType();
}

abstract class UtilsAccessedPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file read" }
}

abstract class UtilsControlledDataSink extends UtilsSink {
  override string sinkType() { result = "tainted data being written to a file" }
}

abstract class UtilsControlledPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file operations" }
}

/**
 * This represents the data in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * {foo:'bar'}
 * ```
 */
class WrittenData extends UtilsControlledDataSink {
  WrittenData() { exists(FileWriters fw | fw.getData() = this) }
}

/**
 * This represents the filepath accessed as an input for the data in calls as follows:
 * ```javascript
 * await copy('db/data').to('dist/db/data')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'db/data'
 * ```
 */
class AccessedPath extends UtilsAccessedPathSink {
  AccessedPath() { exists(FileReaderWriters fw | fw.getFromPath() = this) }
}

/**
 * This represents the filepath where data is written or a file operation is performed in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'some'
 * 'file.json'
 * ```
 */
class ControlledInputPath extends UtilsControlledPathSink {
  ControlledInputPath() {
    exists(FileReaders fw | fw.getPath() = this)
    or
    exists(FileReaderWriters fw | fw.getToPath() = this)
    or
    exists(FileWriters fw | fw.getPath() = this)
    or
    exists(DirectoryWriters dw | dw.getPath() = this)
    or
    exists(DirectoryReaders dr | dr.getPath() = this)
  }
}
