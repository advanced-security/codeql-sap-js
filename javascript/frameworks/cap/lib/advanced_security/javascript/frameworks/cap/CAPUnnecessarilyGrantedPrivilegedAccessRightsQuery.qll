import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A reference to an entity that is defined in this application.
 */
class LocalEntityReference instanceof EntityReference {
  LocalEntityReference() { not this instanceof RemoteEntityReference }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }

  predicate hasRestrictedAccessControl() {
    exists(RestrictCondition restrict |
      restrict =
        this.(EntityReference).getCqlDefinition().getRestrictAnnotation().getARestrictCondition()
    |
      not restrict.grantsToAnyone(_)
    )
  }
}

/**
 * A reference to an entity that is not defined in this application and
 * read from a service instance that is looked up with the name defined in
 * package.json.
 */
class RemoteEntityReference instanceof EntityReference {
  RemoteEntityReference() { not exists(this.getCqlDefinition()) }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class PrivilegedUserInstance extends DataFlow::Node { }

class CdsUserPrivilegedProperty extends PrivilegedUserInstance instanceof PropRead {
  CdsUserPrivilegedProperty() {
    exists(CdsUser cdsUser, PropRead cdsUserRef |
      cdsUserRef = cdsUser.getInducingNode() and
      cdsUserRef.flowsTo(this.getBase()) and
      this.getPropertyName() = "Privileged"
    )
  }
}

class CustomPrivilegedUser extends ClassNode {
  CustomPrivilegedUser() {
    exists(CdsUser cdsUser | this.getASuperClassNode() = cdsUser.asSource()) and
    exists(FunctionNode init |
      init = this.getInstanceMethod("is") and
      forall(Expr expr | expr = init.asExpr().(Function).getAReturnedExpr() |
        expr.mayHaveBooleanValue(true)
      )
    )
  }
}

class CustomPrivilegedUserInstance extends PrivilegedUserInstance, NewNode {
  CustomPrivilegedUserInstance() {
    exists(CustomPrivilegedUser customPrivilegedUserClass |
      this = customPrivilegedUserClass.getAnInstantiation()
    )
  }
}
