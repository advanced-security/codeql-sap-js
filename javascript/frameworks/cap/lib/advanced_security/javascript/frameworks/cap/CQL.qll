import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * Objects from the SQL-like fluent API
 * this is the set of clauses that acts as the base of a statement
 */
class CqlQueryBase extends VarRef {
  CqlQueryBase() {
    exists(string name |
      this.getName() = name and
      name in ["SELECT", "INSERT", "DELETE", "UPDATE", "UPSERT"] and
      (
        /* Made available as a global variable */
        exists(GlobalVariable queryBase | this = queryBase.getAReference())
        or
        /* Imported from `cds.ql` */
        exists(CdsFacade cds |
          cds.getMember("ql").getMember(name).getAValueReachableFromSource().asExpr() = this
        )
      )
    )
  }
}

class CqlSelectBase extends CqlQueryBase {
  CqlSelectBase() { this.getName() = "SELECT" }
}

class CqlInsertBase extends CqlQueryBase {
  CqlInsertBase() { this.getName() = "INSERT" }
}

class CqlDeleteBase extends CqlQueryBase {
  CqlDeleteBase() { this.getName() = "DELETE" }
}

class CqlUpdateBase extends CqlQueryBase {
  CqlUpdateBase() { this.getName() = "UPDATE" }
}

class CqlUpsertBase extends CqlQueryBase {
  CqlUpsertBase() { this.getName() = "UPSERT" }
}

abstract class CqlQueryBaseCall extends CallExpr {
  // TODO: Express "It's a global function or a local function imported from cds.ql"
}

class CqlSelectBaseCall extends CqlQueryBaseCall {
  CqlSelectBaseCall() { this.getCalleeName() = "SELECT" }
}

class CqlInsertBaseCall extends CqlQueryBaseCall {
  CqlInsertBaseCall() { this.getCalleeName() = "INSERT" }
}

/* The cds-ql docs do not mention DELETE being a function acting as a shortcut to any underlying clause. */
class CqlUpdateBaseCall extends CqlQueryBaseCall {
  CqlUpdateBaseCall() { this.getCalleeName() = "UPDATE" }
}

class CqlUpsertBaseCall extends CqlQueryBaseCall {
  CqlUpsertBaseCall() { this.getCalleeName() = "UPSERT" }
}

/**
 * Obtains the receiver across a variety of types of accesses
 */
Expr getRootReceiver(Expr e) {
  result = e and
  (
    e instanceof VarRef
    or
    e instanceof CallExpr and not exists(e.(CallExpr).getReceiver())
  )
  or
  result = getRootReceiver(e.(DotExpr).getBase())
  or
  result = getRootReceiver(e.(MethodCallExpr).getReceiver())
  or
  result = getRootReceiver(e.(PropAccess).getBase())
  or
  result = getRootReceiver(e.(TaggedTemplateExpr).getTag())
}

/**
 * An aggregation type for the two ways to access the fluent API
 * provided by the module cds.ql
 */
private newtype TCqlClause =
  TTaggedTemplate(TaggedTemplateExpr taggedTemplateExpr) {
    exists(CqlQueryBase base | base = getRootReceiver(taggedTemplateExpr)) or
    exists(CqlQueryBaseCall call | call = getRootReceiver(taggedTemplateExpr))
  } or
  TMethodCall(MethodCallExpr callExpr) {
    exists(CqlQueryBase base | base = getRootReceiver(callExpr)) or
    exists(CqlQueryBaseCall call | call = getRootReceiver(callExpr))
  } or
  TShortcutCall(CqlQueryBaseCall callExpr)

class CqlClause extends TCqlClause {
  TaggedTemplateExpr asTaggedTemplate() { this = TTaggedTemplate(result) }

  MethodCallExpr asMethodCall() { this = TMethodCall(result) }

  CallExpr asShortcutCall() { this = TShortcutCall(result) }

  Node flow() { result = this.asExpr().flow() }

  Expr asExpr() {
    result = this.asTaggedTemplate()
    or
    result = this.asMethodCall()
    or
    result = this.asShortcutCall()
  }

  predicate isSelect() {
    /*this.isFinal() and*/
    this.getTypeString() = "SELECT"
  }

  predicate isInsert() { /*this.isFinal() and*/ this.getTypeString() = "INSERT" }

  predicate isUpdate() { /*this.isFinal() and*/ this.getTypeString() = "UPDATE" }

  predicate isUpsert() { /*this.isFinal() and*/ this.getTypeString() = "UPSERT" }

  predicate isDelete() { /*this.isFinal() and*/ this.getTypeString() = "DELETE" }

  predicate isRead() { this.isSelect() }

  predicate isWrite() { this.isInsert() or this.isUpdate() or this.isUpsert() or this.isDelete() }

  Expr getArgument() {
    result = this.asTaggedTemplate().getTemplate() or
    result = this.asMethodCall().getAnArgument() or
    result = this.asShortcutCall().getAnArgument()
  }

  /**
   * Convert this `CqlClause` into a `DotExpr`, i.e.
   * get `` SELECT.from`Table`.where `` when given `` SELECT.from`Table`.where`cond` ``.
   */
  DotExpr asDotExpr() {
    result = this.asTaggedTemplate().getTag() or
    result = this.asMethodCall().getCallee()
  }

  string toString() {
    result = this.asTaggedTemplate().toString() or
    result = this.asMethodCall().toString() or
    result = this.asShortcutCall().toString()
  }

  string getTypeString() {
    result = getRootReceiver(this.asTaggedTemplate()).(VarRef).getName() or
    result = getRootReceiver(this.asTaggedTemplate()).(CallExpr).getCalleeName() or
    result = getRootReceiver(this.asMethodCall()).(VarRef).getName() or
    result = getRootReceiver(this.asMethodCall()).(CallExpr).getCalleeName()
  }

  Location getLocation() {
    result = this.asTaggedTemplate().getLocation() or
    result = this.asMethodCall().getLocation() or
    result = this.asShortcutCall().getLocation()
  }

  CqlQueryBase getCqlBase() { result = getRootReceiver(this.asMethodCall()) }

  CqlQueryBaseCall getCqlBaseCall() {
    result = getRootReceiver(this.asTaggedTemplate()) or
    result = getRootReceiver(this.asMethodCall())
  }

  /** Describes a parent expression relation */
  Expr getParentExpr() {
    result = this.asTaggedTemplate().getParentExpr() or
    result = this.asMethodCall().getParentExpr() or
    result = this.asShortcutCall().getParentExpr()
  }

  /**
   * Possible cases for constructing a chain of clauses:
   *
   * (looking at the terminal clause and its possible parent types as tuples: (this, parent))
   * 1. TMethodCall.TMethodCall
   *     - example `(SELECT.from(Table),  SELECT.from(Table).where("col1='*'"))`
   * 2. TShortcutCall.TMethodCall
   *     - example `(SELECT("col1, col2"), SELECT("col1, col2").from("Table"))`
   *
   * Note that ShortcutCalls cannot be added to any clause chain other than the first position, e.g.
   * ``` javascript
   * SELECT("col1, col2").INSERT(col2)  // Invalid!
   * ```
   */
  CqlClause getParentCqlClause() {
    /* ========== The parent is a shortcut call ========== */
    result.asShortcutCall() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asShortcutCall() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asShortcutCall() = this.asShortcutCall().getParentExpr().getParentExpr() or
    /* ========== The parent is a tagged template ========== */
    result.asTaggedTemplate() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asTaggedTemplate() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asTaggedTemplate() = this.asShortcutCall().getParentExpr().getParentExpr() or
    /* ========== The parent is a method call ========== */
    result.asMethodCall() = this.asTaggedTemplate().getParentExpr().getParentExpr() or
    result.asMethodCall() = this.asMethodCall().getParentExpr().getParentExpr() or
    result.asMethodCall() = this.asShortcutCall().getParentExpr().getParentExpr()
  }

  Expr getAnAncestorExpr() {
    result = this.asTaggedTemplate().getParentExpr+() or
    result = this.asMethodCall().getParentExpr+() or
    result = this.asShortcutCall().getParentExpr+()
  }

  CqlClause getAnAncestorCqlClause() {
    result.asTaggedTemplate() = this.getAnAncestorExpr() or
    result.asMethodCall() = this.getAnAncestorExpr() or
    result.asShortcutCall() = this.getAnAncestorExpr()
  }

  /** Describes a child expression relation */
  Expr getAChildExpr() {
    result = this.asTaggedTemplate().getAChildExpr() or
    result = this.asMethodCall().getAChildExpr() or
    result = this.asShortcutCall().getAChildExpr()
  }

  /**
   * the same chain order logic as `getParentCqlClause` but reversed
   */
  CqlClause getAChildCqlClause() {
    /* ========== The parent is a shortcut call ========== */
    result.asShortcutCall() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asShortcutCall() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asShortcutCall() = this.asShortcutCall().getAChildExpr().getAChildExpr() or
    /* ========== The parent is a tagged template ========== */
    result.asTaggedTemplate() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asTaggedTemplate() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asTaggedTemplate() = this.asShortcutCall().getAChildExpr().getAChildExpr() or
    /* ========== The parent is a method call ========== */
    result.asMethodCall() = this.asTaggedTemplate().getAChildExpr().getAChildExpr() or
    result.asMethodCall() = this.asMethodCall().getAChildExpr().getAChildExpr() or
    result.asMethodCall() = this.asShortcutCall().getAChildExpr().getAChildExpr()
  }

  Expr getADescendantExpr() {
    result = this.asTaggedTemplate().getAChildExpr+() or
    result = this.asMethodCall().getAChildExpr+() or
    result = this.asShortcutCall().getAChildExpr+()
  }

  CqlClause getADescendantCqlClause() {
    result.asTaggedTemplate() = this.getADescendantExpr() or
    result.asMethodCall() = this.getADescendantExpr() or
    result.asShortcutCall() = this.getADescendantExpr()
  }

  predicate isFinal() { not exists(this.getParentCqlClause()) }

  /**
   * Matches the given `CqlClause` to its method/property name, nested at arbitrary depth.
   */
  string getAnAPIName() {
    result = this.asDotExpr().getPropertyName() or
    result = this.getADescendantCqlClause().getAnAPIName()
  }

  abstract CqlClause getEntityAccessingClause();

  /**
   * Gets the reference to the entity that this clause is accessing.
   */
  ExprNode getAccessingEntityReference() {
    result = this.getEntityAccessingClause().getArgument().flow()
  }

  /**
   * Gets the reference of the service that runs this CQL clause.
   */
  ServiceInstance getRunner() {
    exists(CdsTransaction tx | this = tx.getAnExecutedCqlClause() and result = tx.getRunner())
    or
    exists(SrvRun srvRun | this = srvRun.getCql() and result = srvRun.getRecipient())
  }

  CdlEntity getAccessingEntityDefinition() {
    /* 1. String literals or template strings */
    result.getName() =
      this.getAccessingEntityReference().(EntityReferenceFromCqlClause).getStringValue()
    or
    result.getUnqualifiedName() =
      this.getAccessingEntityReference().(EntityReferenceFromCqlClause).getStringValue() and
    /* The entity is accessed by its own service. */
    result = this.getRunner().getDefinition().getCdsDeclaration().getAnEntity()
    or
    /* 2. Variable whose value is a reference to an entity */
    result =
      this.getAccessingEntityReference()
          .getALocalSource()
          .(EntityReferenceFromEntities)
          .getCqlDefinition()
  }
}

class CqlSelectClause extends CqlClause {
  CqlSelectClause() { this.isSelect() }

  override CqlSelectClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "from"
  }
}

class CqlInsertClause extends CqlClause {
  CqlInsertClause() { this.isInsert() }

  override CqlInsertClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "into"
  }
}

class CqlUpdateClause extends CqlClause {
  CqlUpdateClause() { this.isUpdate() }

  override CqlUpdateClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    (
      result.asDotExpr().getPropertyName() = "entity" or
      exists(result.asShortcutCall())
    )
  }
}

class CqlUpsertClause extends CqlClause {
  CqlUpsertClause() { this.isUpsert() }

  override CqlUpsertClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "into"
  }
}

class CqlDeleteClause extends CqlClause {
  CqlDeleteClause() { this.isDelete() }

  override CqlDeleteClause getEntityAccessingClause() {
    result = this.getADescendantCqlClause() and
    result.asDotExpr().getPropertyName() = "from"
  }
}
