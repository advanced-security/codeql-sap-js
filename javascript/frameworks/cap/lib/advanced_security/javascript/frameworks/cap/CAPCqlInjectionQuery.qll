import javascript
import semmle.javascript.security.dataflow.SqlInjectionCustomizations
import advanced_security.javascript.frameworks.cap.CQL
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.dataflow.FlowSteps

class CqlClauseWithStringConcatParameter instanceof CqlClause {
  CqlClauseWithStringConcatParameter() {
    exists(DataFlow::Node queryParameter |
      (
        if this instanceof CqlInsertClause or this instanceof CqlUpsertClause
        then
          queryParameter = this.getArgument().flow() or
          queryParameter = this.getArgument().flow().(SourceNode).getAPropertyWrite().getRhs()
        else queryParameter = this.getArgument().flow()
      ) and
      exists(StringConcatenation::getAnOperand(queryParameter))
    )
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlShortcutMethodCallWithStringConcat instanceof CqlShortcutMethodCall {
  CqlShortcutMethodCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getAQueryParameter()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlClauseParserCallWithStringConcat instanceof CqlClauseParserCall {
  CqlClauseParserCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getCdlString()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlInjectionConfiguration extends TaintTracking::Configuration {
  CqlInjectionConfiguration() { this = "CQL injection from untrusted data" }

  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }

  override predicate isSink(DataFlow::Node node) {
    exists(CqlRunMethodCall cqlRunMethodCall |
      node = cqlRunMethodCall.(CqlRunMethodCall).getAQueryParameter()
    )
    or
    exists(CqlShortcutMethodCallWithStringConcat queryRunnerCall |
      node = queryRunnerCall.(CqlQueryRunnerCall).getAQueryParameter()
    )
    or
    exists(AwaitExpr await, CqlClauseWithStringConcatParameter cqlClauseWithStringConcat |
      node = await.flow() and
      await.getOperand() = cqlClauseWithStringConcat.(CqlClause).asExpr()
    )
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof SqlInjection::Sanitizer }

  override predicate isAdditionalTaintStep(DataFlow::Node start, DataFlow::Node end) {
    exists(CqlClauseParserCallWithStringConcat cqlParseCallWithStringConcat |
      start = cqlParseCallWithStringConcat.(CqlClauseParserCall).getAnArgument() and
      end = cqlParseCallWithStringConcat
    )
    or
    exists(CqlClause cqlClause |
      start = cqlClause.getArgument().flow() and
      end = cqlClause.flow()
    )
  }
}
