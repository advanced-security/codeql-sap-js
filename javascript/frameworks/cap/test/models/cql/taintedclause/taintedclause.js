/* ========== CQL parse usages ========== */
const cds = require('@sap/cds');
let cqn = CQL`SELECT col1, col2, col3 from Table` + userInput //this is actually already captured by taint steps by default
let cqn1 = cds.parse.cql (`SELECT * from Foo`+ userInput) 
let cqn2 = cds.parse.cql (`SELECT * from Foo`) + userInput //not valid
