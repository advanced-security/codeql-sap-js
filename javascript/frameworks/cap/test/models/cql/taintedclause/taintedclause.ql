import javascript
import advanced_security.javascript.frameworks.cap.CQL

from CQL::ParseCQLTaintedClause clause
select clause