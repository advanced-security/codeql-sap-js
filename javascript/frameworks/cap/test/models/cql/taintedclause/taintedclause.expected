| taintedclause.js:4:12:4:57 | cds.par ... rInput) |
